presto-orc/src/main/java/com/facebook/presto/orc/reader/ApacheHiveTimestampDecoder.java

```diff
@@ -36,8 +37,30 @@ public static long decodeTimestamp(long seconds, long serializedNanos, DecodeTim
         if (value < 0 && nanos != 0) {
             value -= options.getUnitsPerSecond();
         }
-        // Truncate nanos to required units (millis / micros) and add to value
-        return value + (nanos / options.getNanosPerUnit());
+        return getValueWithNanos(value, nanos, options.getNanosPerUnit());
+    }
+
+    private static long getSecondsInRequiredUnits(long secondsWithBase, long unitsPerSecond)
+    {
+        try {
+            return Math.multiplyExact(secondsWithBase, unitsPerSecond);
+        }
+        catch (ArithmeticException e) {
+            throw new IllegalArgumentException(String.format("seconds field of timestamp exceeds maximum supported value, secondsWithBase: %s unitsPerSecond: %s.",
```

Comment:
Do not throw IllegalArgumentException, they are harder to track in the errors. Create a new error TimestampOutOfBounds Exception or something like that.

Reply:
done.